Userland/Libraries/LibWeb/Selection/Selection.cpp

```diff
@@ -137,6 +138,9 @@ void Selection::add_range(JS::NonnullGCPtr<DOM::Range> range)
 
     // 3. Set this's range to range by a strong reference (not by making a copy).
     m_range = range;
+
+    if (auto layout_root = m_document->layout_node())
```

Comment:
You could add a `set_range(GcPtr<Range>)` API to centralize this state update that might look cleaner than reaching into the layout tree in every function directly.

Reply:
It seems like other browsers also react visibly if you manipulate the current selection's range programmatically, e.g via: `document.getSelection().getRangeAt(0).setStartOffset(...)`

So I think we need a way for `Range` to notify `Selection` that it mutated..